### ISSUE TITLE
Performing joins on the query containing nested query with FirstOrDefault fails

### ISSUE THREAD

**Issue by kresimirlesic** [linq2db/linq2db#3334]
Hi,

performing joins on the query containing nested query with FirstOrDefault fails.

```
Exception message:
Object reference not set to an instance of an object.

Stack trace:
   at LinqToDB.Linq.Builder.SelectContext.<>c.<ConvertToIndexInternal>b__49_2(ValueTuple`4 context, Int32 n, IBuildContext ctx, Expression ex, Int32 l, Expression _)
   at LinqToDB.Linq.Builder.SelectContext.ProcessMemberAccess[T,TContext](TContext context, Expression expression, MemberExpression levelExpression, Int32 level, Func`7 action)
   at LinqToDB.Linq.Builder.SelectContext.ConvertToIndexInternal(Expression expression, Int32 level, ConvertFlags flags)
   at LinqToDB.Linq.Builder.SelectContext.ConvertToIndex(Expression expression, Int32 level, ConvertFlags flags)
   at LinqToDB.Linq.Builder.SubQueryContext.ConvertToSql(Expression expression, Int32 level, ConvertFlags flags)
   at LinqToDB.Linq.Builder.ExpressionContext.ConvertToSql(Expression expression, Int32 level, ConvertFlags flags)
   at LinqToDB.Linq.Builder.ExpressionBuilder.ConvertToSql(IBuildContext context, Expression expression, Boolean unwrap, ColumnDescriptor columnDescriptor, Boolean isPureExpression)
   at LinqToDB.Linq.Builder.JoinBuilder.BuildJoin(ExpressionBuilder builder, SqlSearchCondition condition, IBuildContext outerKeyContext, Expression outerKeySelector, IBuildContext innerKeyContext, Expression innerKeySelector)
   at LinqToDB.Linq.Builder.JoinBuilder.BuildMethodCall(ExpressionBuilder builder, MethodCallExpression methodCall, BuildInfo buildInfo)
   at LinqToDB.Linq.Builder.MethodCallBuilder.BuildSequence(ExpressionBuilder builder, BuildInfo buildInfo)
   at LinqToDB.Linq.Builder.ExpressionBuilder.BuildSequence(BuildInfo buildInfo)
   at LinqToDB.Linq.Builder.FirstSingleBuilder.BuildMethodCall(ExpressionBuilder builder, MethodCallExpression methodCall, BuildInfo buildInfo)
   at LinqToDB.Linq.Builder.MethodCallBuilder.BuildSequence(ExpressionBuilder builder, BuildInfo buildInfo)
   at LinqToDB.Linq.Builder.ExpressionBuilder.BuildSequence(BuildInfo buildInfo)
   at LinqToDB.Linq.Builder.ExpressionBuilder.Build[T]()
   at LinqToDB.Linq.Query`1.CreateQuery(ExpressionTreeOptimizationContext optimizationContext, ParametersContext parametersContext, IDataContext dataContext, Expression expr)
   at LinqToDB.Linq.Query`1.GetQuery(IDataContext dataContext, Expression& expr, Boolean& dependsOnParameters)
   at LinqToDB.Linq.ExpressionQuery`1.GetQuery(Expression& expression, Boolean cache, Boolean& dependsOnParameters)
   at LinqToDB.Linq.ExpressionQuery`1.System.Linq.IQueryProvider.Execute[TResult](Expression expression)
   at System.Linq.Queryable.FirstOrDefault[TSource](IQueryable`1 source)
   at Linq2dbTest.Program.Main(String[] args) in Program.cs:line 34

```

### Steps to reproduce

```
using LinqToDB;
using LinqToDB.Mapping;
using System.Linq;

namespace Linq2dbTest
{
    public static class Program
    {
        public static void Main()
        {
            var builder = MappingSchema.Default.GetFluentMappingBuilder();

            builder.Entity<Entity1>();

            builder.Entity<Entity2>();

            builder.Entity<Entity3>();

            using (var con = new Db())
            {
                var subquery = con.GetTable<Entity2>();

                var query = con.GetTable<Entity1>()
                    .Select(entity1 => new
                    {
                        Entity1 = entity1,
                        Entity2 = subquery.FirstOrDefault(entity2 => entity2.Id == entity1.Id)
                    })
                    .GroupJoin(con.GetTable<Entity3>(),
                        x => x.Entity2.Id,
                        x => x.Id,
                        (x, y) => x);

                var result = query.FirstOrDefault();
            }
        }

        public class Db : LinqToDB.Data.DataConnection
        {
            public Db() : base(ProviderName.SqlServer2017,
                "data source=localhost;initial catalog=xxx;integrated security=True;MultipleActiveResultSets=True")
            {
            }
        }

        public class Entity1
        {
            public int Id { get; set; }
        }

        public class Entity2
        {
            public int Id { get; set; }
        }

        public class Entity3
        {
            public int Id { get; set; }
        }
    }
}
```

### Environment details
linq2db version: 3.5.1
Database Server: SQL Server 2019
Database Provider: SQL
Operating system: W10
.NET Framework: 4.8

**Comment by sdanyliv** [comment]
@kresimirlesic, do you need workaround? `GroupJoin` is not SQL friendly operator.

**Comment by kresimirlesic** [comment]
This is a really simplified and stripped down version of the actual query.

What I need is to be able to do multiple stuff with the results of `Entity2 = subquery.FirstOrDefault(entity2 => entity2.Id == entity1.Id)`.

That includes:
1. left joining on the Entity2 and using the results of that join in the upcoming select (kinda like in the example above, but obviously the select many and select are missing)
2. being able to use the navigation properties on Entity2 in the upcoming select (this is not in the example above)

If you have any ideas about the possible workarounds that will make the above work, feel free to share them and I can try if they work.

**Comment by kresimirlesic** [comment]
@sdanyliv hey, any updates on this one?

**Comment by stomasovic** [comment]
Hi, 

We succeeded using Where.Take.DefaultIfEmpty instead of FirstOrDefault
`
var query = con.GetTable<Entity1>()
                    .Select(entity1 => new
                    {
                        Entity1 = entity1,
                        Entity2 = subquery.FirstOrDefault(entity2 => entity2.Id == entity1.Id)
                    })
`
use following instead:
`
var query = con.GetTable<Entity1>()
                    .Select(entity1 => new
                    {
                        Entity1 = entity1,
                        Entity2 = subquery.Where(entity2 => entity2.Id == entity1.Id).Take(1).DefaultIfEmpty()
                    })
`

**Comment by sdanyliv** [comment]
Sorry for delay, try this one. Query syntax is more preferable here:
```cs
var query = 
    from e1 in con.GetTable<Entity1>()
    from e2 in query.Where(e2 => e2.Id == e1.Id)
        .Take(1)
        .DefaultIfEmpty()
    from e3 in con.GetTable<Entity3>().Where(e3 => e3.Id == e2.Id)
    select e3;
```
I'm working on fixing FirstOrDefault in queries, note easy task.

**Comment by MaceWindu** [comment]
Fixed in `6.0.0-preview.1`
